### PR TITLE
Introduce http & http2 options with timeouts

### PR DESCRIPTION
**Pull request by felixputera**
Addresses #68 

The general idea follows the discussion from the issue: introduce `http` & `http2` options, along with timeouts handling.

The new API:

```js
fastify.register(require('fastify-reply-from'), {
  http: {
    agentOptions: { // supports all options from https://nodejs.org/api/https.html#https_new_agent_options
      keepAliveMsecs: 30000, // supersedes options.keepAliveMsecs
      maxFreeSockets: 100, // supersedes options.maxFreeSockets
      maxSockets: 200 // supersedes options.maxSockets
    },
    requestOptions : { // supports all options from https://nodejs.org/api/https.html#https_https_request_url_options_callback
      timeout: 5000, // has default of 10000, with attached timeout callback
      rejectUnauthorized: true // supersedes options.rejectUnauthorized, defaults to false (Node defaults to true)
    }
  },
  http2: {
    sessionTimeout: 30000, // supersedes options.sessionTimeout, now with attached timeout callback
    requestTimeout: 5000, // has default of 10000, with attached timeout callback
    sessionOptions: { // supports all options from https://nodejs.org/api/http2.html#http2_http2_connect_authority_options_listener
       rejectUnauthorized: true // supersedes options.rejectUnauthorized, defaults to false (Node defaults to true)
    },
    requestOptions : { // supports all options from https://nodejs.org/api/http2.html#http2_clienthttp2session_request_headers_options
      endStream: true // just some random request.options as an illustration :)
    }
  }
}
```

Would love to get some feedbacks before continuing with the rest of TODOs.

TODO:
- [X] update docs (new readme is accessible at https://github.com/felixputera/fastify-reply-from for rich preview)
- [X] TS declarations tests

#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [X] documentation is changed or added
- [X] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -102,6 +105,8 @@ module.exports = fp(function from (fastify, opts, next) {
         if (!this.sent) {
           if (err.code === 'ERR_HTTP2_STREAM_CANCEL') {
             this.code(503).send(new Error('Service Unavailable'))
+          } else if (err.code === ERR_TIMED_OUT_CODE) {
+            this.code(504).send('Gateway Timeout')
```
Comment (jsumners) [inline index.js:109]: This should be an error instance.

Comment (felixputera) [reply index.js:109]: I am currently throwing bunch of different error instances with different error messages in `lib/request.js`. I initially planned to throw the errors to the library users, but with the current approach you suggestion makes a lot of sense 👍.

Comment (jsumners) [reply index.js:109]: See line 107. Why is line 109 only returning a string?

**Comment by felixputera** [comment]
@mcollina I had added the tests & docs, please help me review them too!

File: README.md [hunk 2]
```diff
@@ -59,17 +59,65 @@ target.listen(3001, (err) => {
...
 ## API
 
-### plugin options
+### Timeout
+
+This library has:
+- `timeout` for `http` set by default. The default value is 10 seconds (`10000`).
+- `requestTimeout` & `sessionTimeout` for `http2` set by default.
+  - The default value for `requestTimeout` is 10 seconds (`10000`).
+  - The default value for `requestTimeout` is 60 seconds (`60000`).
```
Comment (adrai) [inline README.md:68]: probably sessionTimeout?

Comment (felixputera) [reply README.md:68]: ah yes 😅

File: README.md [hunk 3]
```diff
@@ -59,17 +59,65 @@ target.listen(3001, (err) => {
...
+
+When a timeout happens, [`504 Gateway Timeout`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/504)
+will be returned to the client.
```
[the remark this answers is not in the document]
Comment (felixputera) [reply README.md:71]: I am unsure of where to put this actually. Do you have a suggestion under **Plugin options** more precisely where we should put this?

[the remark this answers is not in the document]
Comment (mcollina) [reply README.md:71]: Just move it as is after plugin options.

File: README.md [hunk 4]
```diff
@@ -86,76 +123,95 @@ proxy.register(require('fastify-reply-from'), {
...
+#### `maxFreeSockets`
 
-#### rejectUnauthorized
+**(Deprecated)** Defaults to `2048` free sockets, passed down to [`http.Agent`][http-agent] and
+[`https.Agent`][https-agent] instances. Prefer to use [`http.agentOptions`](#http) instead.
 
-Defaults to `true`, passed down to [`https.Agent`][https-agent] instances.
+#### `rejectUnauthorized`
+
+**(Deprecated)** Defaults to `false`, passed down to [`https.Agent`][https-agent] instances.
```
Comment (Eomm) [inline README.md:147]: Should not emit a warning when deprecated features are used?

File: README.md [hunk 5]
```diff
@@ -210,6 +199,19 @@ Setting this option will not verify if the http method allows for a body.
 Override the `'Content-Type'` header of the forwarded request, if we are
 already overriding the [`body`][body].
 
+---
```
Comment (mcollina) [inline README.md:202]:
```suggestion
```

File: README.md [hunk 6]
```diff
@@ -210,6 +199,19 @@ Setting this option will not verify if the http method allows for a body.
...
+
+#### Timeout
```
Comment (mcollina) [inline README.md:204]:
```suggestion
### HTTP & HTTP2 timeouts
```

File: README.md (hunk 4 above)
Comment (felixputera) [reply README.md:147]: Sorry, I didn't really get what you mean there. 

Do you mean we should add some deprecation warning in code? Or add some docs documenting that the library would not emit a warning when a deprecated feature is used?

Comment (Eomm) [reply README.md:147]: Yes, I mean to add a (`once`) warning in the code that the developer will see:

https://github.com/fastify/fastify/blob/master/lib/warnings.js

Comment (felixputera) [reply README.md:147]: Gotcha. Will do so!

Comment (Eomm) [reply README.md:147]: If you could do it in a following PR it would be appreciated since this PR has been merged 👍

**Pull request merged**